File: app/src/main/java/com/upclicks/ffcapp/viewModels/AccountViewModel.java
package com.upclicks.ffcapp.viewModels;

import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import com.upclicks.ffcapp.base.BaseViewModel;
import com.upclicks.ffcapp.data.CustomRxObserver;
import com.upclicks.ffcapp.models.Result;
import com.upclicks.ffcapp.models.requests.LoginRequest;
import com.upclicks.ffcapp.models.requests.SignUpRequest;
import com.upclicks.ffcapp.models.requests.UpdateProfileRequest;
import com.upclicks.ffcapp.models.response.City;
import com.upclicks.ffcapp.models.response.Profile;
import com.upclicks.ffcapp.repositories.AccountRepository;
import com.upclicks.ffcapp.session.UserSession;

import java.util.HashMap;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.schedulers.Schedulers;
import okhttp3.MultipartBody;

public class AccountViewModel extends BaseViewModel {
    private AccountRepository accountRepository;
    private MutableLiveData<String> updateProfileLive = new MutableLiveData<>();
    private MutableLiveData<UserSession> userSessionMutableLiveData = new MutableLiveData<>();
    private MutableLiveData<List<City>> citesLiveData = new MutableLiveData<>();
    private MutableLiveData<Profile> userProfile = new MutableLiveData<>();
    private MutableLiveData<String> avatarUpdateRes = new MutableLiveData<>();
    private MutableLiveData<UserSession> changingPasswordResponse = new MutableLiveData<>();

    @ViewModelInject
    public AccountViewModel(AccountRepository accountRepository) {
        this.accountRepository = accountRepository;
    }

    public LiveData<UserSession> observeLogin(){
        return userSessionMutableLiveData;
    }
    public LiveData<UserSession> observeSignUp(){
        return userSessionMutableLiveData;
    }
    public LiveData<List<City>> observeCites(){return citesLiveData;}
    public LiveData<String> observeUpdateProfile (){
        return updateProfileLive;
    }
    public LiveData<Profile> observeUserProfile(){return userProfile;}
    public LiveData<String> observeUpdatingAvatar(){return avatarUpdateRes;}
    public LiveData<UserSession> observeChangingPassword(){
        return changingPasswordResponse;
    }
    //Login
    public void login(LoginRequest loginRequest){
        loading.postValue(true);
        accountRepository.login(loginRequest)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new CustomRxObserver(this) {
                    @Override
                    public void onResponse(Result<Object> response) {
                        userSessionMutableLiveData.postValue((UserSession) response.getResult());
                    }
                });
    }
    //Sign Up
    public void signUp(SignUpRequest signUpRequest){
        loading.postValue(true);
        accountRepository.signUp(signUpRequest)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new CustomRxObserver(this) {
                    @Override
                    public void onResponse(Result<Object> response) {
                        userSessionMutableLiveData.postValue((UserSession) response.getResult());
                    }
                });
    }
    //Get Profile
    public void getUserProfile(){
        loading.postValue(true);
        accountRepository.getUserProfile()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new CustomRxObserver(this) {
                    @Override
                    public void onResponse(Result<Object> response) {
                        userProfile.postValue((Profile) response.getResult());
                    }
                });
    }
    //Get Cites
    public void getCites(){
        accountRepository.getCites()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new CustomRxObserver(this) {
                    @Override
                    public void onResponse(Result<Object> response) {
                        citesLiveData.postValue((List<City>) response.getResult());
                    }
                });
    }
    //Update Profile
    public void updateProfile(UpdateProfileRequest updateProfileRequest){
        loading.postValue(true);
        accountRepository.updateProfile(updateProfileRequest)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new CustomRxObserver(this) {
                    @Override
                    public void onResponse(Result<Object> response) {
                        updateProfileLive.postValue((String) response.getResult());
                    }
                });
    }
    //Update avatar
    public void updateAvatar(MultipartBody.Part avatar){
        loading.postValue(true);
        accountRepository.updateAvatar(avatar)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new CustomRxObserver(this) {
                    @Override
                    public void onResponse(Result<Object> response) {
                        avatarUpdateRes.postValue((String) response.getResult());
                    }
                });
    }
    //Change password
    public void changePassword(String currentPassword , String newPassword){
        HashMap request = new HashMap();
        request.put("CurrentPassword",currentPassword);
        request.put("NewPassword",newPassword);
        loading.postValue(true);
        accountRepository.changePassword(request)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new CustomRxObserver(this) {
                    @Override
                    public void onResponse(Result<Object> response) {
                        changingPasswordResponse.postValue((UserSession) response.getResult());
                    }
                });
    }
}

File: app/src/main/java/com/upclicks/ffc/ui/authentication/viewmodel/AccountViewModel.kt
package com.upclicks.ffc.ui.authentication.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.upclicks.ffc.base.BaseViewModel
import com.upclicks.ffc.ui.authentication.model.response.*
import com.upclicks.ffc.rx.CustomRxObserver
import com.upclicks.ffc.ui.authentication.model.Session
import com.upclicks.ffc.ui.authentication.repositories.AccountRepository
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.schedulers.Schedulers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import okhttp3.MultipartBody
import javax.inject.Inject
import com.upclicks.ffc.ui.authentication.model.request.*
import com.upclicks.ffc.ui.general.model.City
import com.upclicks.ffc.ui.general.model.Country
import com.upclicks.ffc.ui.general.model.FeedbackRequest
import com.upclicks.ffc.data.remote.Result
import dagger.hilt.android.lifecycle.HiltViewModel


@ExperimentalCoroutinesApi
@HiltViewModel
class AccountViewModel
@Inject constructor(
    private val accountRepository: AccountRepository,
) : BaseViewModel() {
    private val _authResponse: MutableLiveData<Session> = MutableLiveData()
    private val _verifySessionResponse: MutableLiveData<VerifySession> = MutableLiveData()

    val authResponse: LiveData<Session>
        get() = _authResponse
    val observeVerifySession: LiveData<VerifySession>
        get() = _verifySessionResponse

    fun auth(loginRequest : LoginRequest) {
        loginRequest.rememberClient = true
        accountRepository.login(loginRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<Session>>(this@AccountViewModel) {
                override fun onResponse(response: Result<Session>) {
                    _authResponse.postValue(response!!.result)
                }
            })
    }


    //call Verify Session
    fun callVerifySession() {
        accountRepository.verifySession()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<VerifySession>>(this@AccountViewModel) {
                override fun onResponse(response: Result<VerifySession>) {
                    _verifySessionResponse?.postValue(response.result)
                }
            })
    }


    //Get Countries
    fun getCounties(onGetCountries: (List<Country>) -> Unit) {
        accountRepository.getCountries()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<List<Country>>>(this@AccountViewModel) {
                override fun onResponse(response: Result<List<Country>>) {
                    onGetCountries(response?.result!!)
                }
            })
    }

    //Get Cities
    fun getCities(onGetCites: (List<City>) -> Unit) {
        accountRepository.getCities()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<List<City>>>(this@AccountViewModel) {
                override fun onResponse(response: Result<List<City>>) {
                    onGetCites(response.result!!)
                }
            })
    }

    //Create membership
    fun createMembershipRequest(
        memberShipRequest: CreateMemberShipRequest,
        onMembershipRequestCreated: (MembershipResponse) -> Unit
    ) {
        accountRepository.createMembership(memberShipRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object :
                CustomRxObserver<Result<MembershipResponse>>(this@AccountViewModel) {
                override fun onResponse(response: Result<MembershipResponse>) {
                    onMembershipRequestCreated(response?.result!!)
                }
            })
    }

    //sendFeedback
    fun sendFeedback(
        feedbackRequest: FeedbackRequest,
        onResult: (String) -> Unit
    ) {
        accountRepository.sendFeedback(feedbackRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object :
                CustomRxObserver<Result<String>>(this@AccountViewModel) {
                override fun onResponse(response: Result<String>) {
                    onResult(response?.result!!)
                }
            })
    }

    //Verify membership
    fun verifyMembership(
        membershipIdentity: String,
        verificationCode: String,
        onMembershipVerified: (Boolean) -> Unit
    ) {
        accountRepository.verifyMembership(membershipIdentity, verificationCode)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<Boolean>>(this@AccountViewModel) {
                override fun onResponse(response: Result<Boolean>) {
                    onMembershipVerified(response?.result!!)
                }
            })
    }


    //Complete profile
    fun completeProfile(
        completeProfileRequest: CompleteProfileRequest,
        onProfileCompleted: (String) -> Unit
    ) {
        accountRepository.completeProfile(completeProfileRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<String>>(this@AccountViewModel) {
                override fun onResponse(response: Result<String>) {
                    onProfileCompleted(response?.result!!)
                }
            })
    }


    //Update Avatar
    fun updateAvatar(avatar: MultipartBody.Part, onAvatarUpdated: (String) -> Unit) {
        accountRepository.updateAvatar(avatar)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<String>>(this@AccountViewModel) {
                override fun onResponse(response: Result<String>) {
                    onAvatarUpdated(response?.result!!)
                }
            })
    }


    //Update Profile
    fun updateProfile(updateRequest: UpdateProfileRequest, onProfileUpdated: (String) -> Unit) {
        accountRepository.updateProfile(updateRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<String>>(this@AccountViewModel) {
                override fun onResponse(response: Result<String>) {
                    onProfileUpdated(response?.result!!)
                }
            })
    }


    //Verify Account Identity
    fun verifyAccountIdentity(verifyIDRequest: VerifyIDRequest, onRequestSent: (String) -> Unit) {
        accountRepository.verifyAccountIdentity(verifyIDRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<String>>(this@AccountViewModel) {
                override fun onResponse(response: Result<String>) {
                    onRequestSent(response?.result!!)
                }
            })
    }

    //Get Identity verification status
    fun getIdentityVerificationStatus(
        notifyId: String,
        onStatusLoaded: (IdentityVerificationStatus) -> Unit
    ) {
        accountRepository.getIdentityVerificationStatus(notifyId)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object :
                CustomRxObserver<Result<IdentityVerificationStatus>>(this@AccountViewModel) {
                override fun onResponse(response: Result<IdentityVerificationStatus>) {
                    onStatusLoaded(response?.result!!)
                }
            })
    }

    //Get My Profile
    fun getMyProfile(onMyProfileLoaded: (VerifySession.Profile) -> Unit) {
        accountRepository.getMyProfile()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object :
                CustomRxObserver<Result<VerifySession.Profile>>(this@AccountViewModel) {
                override fun onResponse(response: Result<VerifySession.Profile>) {
                    onMyProfileLoaded(response?.result!!)
                }
            })
    }

    //Change Password
    fun changePassword(
        changePasswordRequest: ChangePasswordRequest,
        onPasswordChanged: (Boolean) -> Unit
    ) {
        accountRepository.changePassword(changePasswordRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<Boolean>>(this@AccountViewModel) {
                override fun onResponse(response: Result<Boolean>) {
                    onPasswordChanged(response?.result!!)
                }
            })
    }

    //Forgot password
    fun forgotPassword(
        emailAddress: String,
        onForgotPasswordRequested: (ForgotPasswordResponse) -> Unit
    ) {
        accountRepository.forgotPassword(emailAddress)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object :
                CustomRxObserver<Result<ForgotPasswordResponse>>(this@AccountViewModel) {
                override fun onResponse(response: Result<ForgotPasswordResponse>) {
                    onForgotPasswordRequested(response?.result!!)
                }
            })
    }

    //Reset password
    fun resetPassword(
        resetPasswordRequest: ResetPasswordRequest,
        onPasswordReseated: (Boolean) -> Unit
    ) {
        accountRepository.resetPassword(resetPasswordRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : CustomRxObserver<Result<Boolean>>(this@AccountViewModel) {
                override fun onResponse(response: Result<Boolean>) {
                    onPasswordReseated(response?.result!!)
                }
            })
    }

}